datalad_next/runners/run.py

```diff
@@ -0,0 +1,166 @@
+"""
+This module provides a context-manager that executes a subprocess.
+
+The context manager supports the termination and/or killing of the subprocess,
+if it is executed with a not-None `terminate_time` and/or `kill_time`.
+
+The context manager supports generator- and non-generator-protocols.
+
+"""
+from __future__ import annotations
+
+import subprocess
+from collections.abc import Generator
+from contextlib import contextmanager
+from pathlib import Path
+from queue import Queue
+from subprocess import DEVNULL
+from typing import (
+    Any,
+    IO,
+)
+
+from datalad_next.runners import (
+    GeneratorMixIn,
+    Protocol,
+    ThreadedRunner,
+)
+
+
+def create_kill_wrapper(cls: type[Protocol]) -> type[Protocol]:
+    """ Extend `cls` to supports the "kill-interface"
+
+    This function creates a subclass of `cls` that contains the components
+    of the "kill-interface".  The two major components are amethod called `arm`,
+    and logic inside the timeout handler that can trigger a termination or
+    kill signal to the subprocess, if the termination time or kill time has
+    come.
+
+    Parameters
+    ----------
+    cls : type[Protocol]
+        A protocol class that should be extended by the kill-interface
+
+    Returns
+    -------
+    Any
+        If the `cls` is not a subclass of `GeneratorMixIn`, the result of
+        `cls._prepare_result` will be returned.
```

Comment:
Do you know why such an essential functionality is provided by a private method?

Reply:
No idea. It was like then when I met Datalad